pkg/apis/kf/v1alpha1/app_validation.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the License);
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an AS IS BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"context"
+
+	"github.com/knative/serving/pkg/apis/serving"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/apis"
+)
+
+// Validate checks for errors in the App's spec or status fields.
+func (app *App) Validate(ctx context.Context) (errs *apis.FieldError) {
+	// If we're specifically updating status, don't reject the change because
+	// of a spec issue.
+	if !apis.IsInStatusUpdate(ctx) {
+		errs = errs.Also(app.Spec.Validate(apis.WithinSpec(ctx)).ViaField("spec"))
+	}
+
+	return errs
+}
+
+// Validate checks that the pod template the user has submitted is valid
+// and that the scaling and lifecycle is valid.
+func (spec *AppSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+
+	errs = errs.Also(ValidatePodSpec(spec.Template.Spec).ViaField("template.spec"))
+	errs = errs.Also(spec.Instances.Validate(ctx).ViaField("instances"))
+
+	return errs
+}
+
+// Validate checks that the fields the user has specified in AppSpecInstances
+// can be used together.
+func (instances *AppSpecInstances) Validate(ctx context.Context) (errs *apis.FieldError) {
+	hasExactly := instances.Exactly != nil
+	hasMin := instances.Min != nil
+	hasMax := instances.Max != nil
+
+	if hasExactly && hasMin {
+		errs = errs.Also(apis.ErrMultipleOneOf("exactly", "min"))
+	}
+
+	if hasExactly && hasMax {
+		errs = errs.Also(apis.ErrMultipleOneOf("exactly", "max"))
+	}
+
+	if hasExactly && *instances.Exactly < 0 {
+		errs = errs.Also(apis.ErrInvalidValue(*instances.Exactly, "exactly"))
+	}
+
+	if hasMin && *instances.Min < 0 {
+		errs = errs.Also(apis.ErrInvalidValue(*instances.Min, "min"))
+	}
+
+	if hasMax && *instances.Max < 0 {
+		errs = errs.Also(apis.ErrInvalidValue(*instances.Max, "max"))
+	}
+
+	if hasMin && hasMax && *instances.Min > *instances.Max {
```

Comment:
note: it's possible that min and max can be the same value. Should we
1. Allow this
2. Allow this but edit the spec to use exactly
3. Explicitly fail on this case, with an error message to use exactly.

Honestly I think what's here, aka 1. is fine, it makes it easy to pro grammatically adjust min/max. But I wanted to bring it up.